File: Algorithm/C++/155. 最小栈.cpp
class MinStack {
private:
    stack<int> my_stack, min_stack;

public:
    MinStack() {
        min_stack.push(INT_MAX);
    }
    
    void push(int val) {
        my_stack.push(val);
        min_stack.push(min(val, min_stack.top()));
    }
    
    void pop() {
        my_stack.pop();
        min_stack.pop();
    }
    
    int top() {
        return my_stack.top();
    }
    
    int getMin() {
        return min_stack.top();
    }
};

File: Algorithm/JavaScript/155. 最小栈.ts
class MinStack {
    protected dataStack: number[] = [];
    protected minStack: number[] = [];

    constructor() { }

    push(val: number): void {
        this.dataStack.push(val)
        if (this.minStack.length === 0) {
            this.minStack.push(val)
        } else {
            this.minStack.push(Math.min(this.getMin(), val))
        }
    }

    pop(): void {
        if (this.dataStack.length === 0) {
            return
        }
        this.dataStack.pop()
        this.minStack.pop()
    }

    top(): number {
        if (this.dataStack.length === 0) {
            return Infinity
        }
        return this.dataStack[this.dataStack.length - 1]
    }

    getMin(): number {
        if (this.minStack.length === 0) {
            return Infinity
        }
        return this.minStack[this.minStack.length - 1]
    }
}